### PR TITLE
Merge sdk2 into develop

### PR DESCRIPTION
**Pull request by ebuchman**

### CODE REVIEW

File: keys/keybase.go [hunk 1]
```diff
@@ -32,58 +26,59 @@ func New(db dbm.DB, codec words.Codec) dbKeybase {
...
+		return Info{}, "", err
 	}
 
 	// encrypt and persist the key
-	public := kb.writeKey(key, name, passphrase)
+	info := kb.writeKey(priv, name, passphrase)
+
+	// we append the type byte to the serialized secret to help with
+	// recovery
+	// ie [secret] = [type] + [secret]
```
Comment (ebuchman) [inline keys/keybase.go:49]: why do this? it seems wrong to influence the words in any way. Why not just pass in the algo explicitly ? This seems like unecessary unstandard weirdness ...

File: keys/keybase.go [hunk 2]
```diff
@@ -32,58 +26,59 @@ func New(db dbm.DB, codec words.Codec) dbKeybase {
...
-	public := kb.writeKey(key, name, passphrase)
-	return public, nil
+	seed := strings.Join(words, " ")
+	return info, seed, err
 }
 
-// SeedToPrivKey returns the private key corresponding to a seedphrase
-// without persisting the private key.
-// TODO: enable the keybase to just hold these in memory so we can sign without persisting (?)
-func (kb dbKeybase) SeedToPrivKey(algo, seedphrase string) (crypto.PrivKey, error) {
```
Comment (ebuchman) [inline keys/keybase.go:75]: What if we just want to use a private key from seed without persisting it to disk? that's what this function is for ...

File: keys/keybase.go [hunk 3]
```diff
@@ -92,156 +87,116 @@ func (kb dbKeybase) List() ([]Info, error) {
...
+	bz := kb.db.Get(infoKey(name))
+	if bz == nil {
+		return "", errors.New("No key to export with name " + name)
 	}
-	armorBytes := encryptArmorPrivKey(key, transferpass)
-	return []byte(armorBytes), nil
+	return armorInfoBytes(bz), nil
 }
 
-// Import accepts bytes generated by Export along with the same transferpass.
```
Comment (ebuchman) [inline keys/keybase.go:149]: Why remove comments?

File: keys/keybase.go (hunk 1 above)
Comment (jaekwon) [reply keys/keybase.go:49]: it's a UX thing.  I don't want to have to remember what the words were for.

Comment (jaekwon) [reply keys/keybase.go:49]: It'll just confused the user.  Maybe we can just prepend a single word that represents the type of key, so it's 13 words?

File: keys/keybase.go (hunk 2 above)
Comment (jaekwon) [reply keys/keybase.go:75]: who calls it?  I couldn't find any usage needs.

File: keys/keybase.go (hunk 3 above)
Comment (jaekwon) [reply keys/keybase.go:149]: The old comment was wrong, and the API I think is self explanatory for now...

File: keys/keybase.go (hunk 2 above)
Comment (ebuchman) [reply keys/keybase.go:75]: We haven't exposed it yet but we should. Also the UI is now asking for it (https://github.com/cosmos/cosmos-sdk/issues/301) . We defintely should have a way to enter a seed and use the key without it being persisted to disk.

File: keys/keybase.go (hunk 3 above)
Comment (ebuchman) [reply keys/keybase.go:149]: Should we not by convention have a proper godoc comment on every exposed item ? Even if it's obvious?

File: keys/keybase.go (hunk 1 above)
Comment (ebuchman) [reply keys/keybase.go:49]: lets do eg. `ed25519{ house table chair ... }`

**Pull request merged**